### PR TITLE
Test class for `array.h` classes

### LINKED ISSUE #136: Add/extend unit tests for classes in field.h and arrays.h
Field has some tests, but these can probably be extended. Also need to extend to test functionality in arrays.h

This will be important to ensure we don't break anything when replacing MATLAB.
In particular: allocation and deallocation of instances of these classes.

### Basic Types
- [x] #152
- [x] #153
- [x] [`MaterialCollection`](https://github.com/UCL/TDMS/blob/0f00b6ecb53c81ac8cad1f396b77852c971dbc48/tdms/include/arrays.h#L89-L92) -> Doesn't actually have any attributes, and only has _one_ protected method `init_xyz_vectors`. Ergo, doesn't have it's own test file.
- [x] #158
- [x] #154
- [x] #155
- [x] #156

### Classes that have base classes as attributes
- [x] [`CCollectionBase`](https://github.com/UCL/TDMS/blob/0f00b6ecb53c81ac8cad1f396b77852c971dbc48/tdms/include/arrays.h#L94-L99) -> Doesn't have a constructor or any methods, and just stores three `XYZVectors` as attributes.
- [x] [`DCollectionBase`](https://github.com/UCL/TDMS/blob/0f00b6ecb53c81ac8cad1f396b77852c971dbc48/tdms/include/arrays.h#L119-L123) -> Doesn't have a constructor or any methods, and just stores two `XYZVectors` as attributes.
- [x] #161
- [x] #162
- [x] #163
- [x] #166
- [x] #173 
- [x] #167

### Subclasses Types
- [x] #178
- [x] #184
- [x] #179 
- [x] #185
- [x] #174
- [x] #177
- [x] #175
- [x] #172 
- [x] #170
- [x] `CCoefficientMatrix` (`Matrix<double>`) - No additional functionality on top of the base class.
- [x] #176

### Test classes for cleanup

- [ ] #165 

## `Field.h` classes

- [x] Expand existing tests, add Docstrings where possible, add unit tests that are missing

### LINKED ISSUE #165: `array.h` unit test cleanup
A number of the classes in `arrays.h` use functions from `matlabio.h` in their constructors (or `initialise()` methods, which perform essentially identical setup roles).

Currently, #136 is introducing unit tests for the classes in `arrays.h`, which in turn are performing identical checks against the functions the various class methods call in `matlabio.h`. It would be more prudent to create proper unit tests for functions in `matlabio.h` (in particular, the various `assert_` functions), which in turn will reduce the size of the array unit test files and remove some code duplication.

### PR DESCRIPTION
**Pull request by willGraham01**
Closes #165 | Hence also closes #136 |

# TLDR;

- (~200 lines) Most important introduction (`AbstractArrayTest` class): [LINK](https://github.com/UCL/TDMS/blob/wgraham-array_test_class/tdms/tests/include/abstract_array_test_class.h)
- (~300 lines) A (very boring) long list of test class declarations: [LINK](https://github.com/UCL/TDMS/blob/wgraham-array_test_class/tdms/tests/include/array_test_class.h)
- Everything else is a refactor

---

## `AbstractArrayTest` class

Introduces the `AbstractArrayTest` class, using the template described [here](https://github.com/UCL/TDMS/pull/159#pullrequestreview-1178788597).

This class contains a number of virtual methods, which are designed to be overwritten by the subclasses that correspond to the individual arrays in `arrays.h`. Commonly executed tasks, like creation of MATLAB inputs, have been given explicit `test_{name}`s to save on code repetition. Additionally, the `matlab_input_assigned` variable exists to keep track of whether there is allocated MATLAB memory to clean up, which is picked up at the end of each test.

The only non-virtual method is `run_all_class_tests()`, which does what you think. This method calls all of the `test_` methods that have been overwritten in the subclasses. _(Actually, it technically calls the non-overwritten function method, which in turn just flags the test to be skipped with no checks performed)._ This function also makes use of the `matlab_input_assigned` variable to avoid memory leaks, by forcing `MATLAB` memory to be freed if it has been assigned.

### Sublcasses for each test

Subclasses are created for each custom class defined in `arrays.h`. Subclasses have the `test_` methods overridden to their specific needs, and with `test_` methods that are not overwritten being (effectively) skipped.

The PR's large diff is mostly because of the refactor this requires plus standardising several variable names across multiple classes.

### CODE REVIEW

**Review by samcunliffe — APPROVED** [review]
Sorry for leaving this hanging. All looks good. I still see #191.

File: tdms/tests/include/abstract_array_test_class.h [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+   */
+  bool destroy_matlab_input() {
+    if (matlab_input_assigned) {
+      mxDestroyArray(matlab_input);
+      matlab_input_assigned = false;
+      return true;
+    } else {
+      return false;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply tdms/tests/include/abstract_array_test_class.h:47]: I debated it, but ultimately I like having the `bool` because:
a) It forces me to actually check that I've freed MATLAB memory
b) It forces me to be aware that _some_ MATLAB functions do cleanup after themselves
c) It's a good indicator for where we're going to have to be careful with free-ing memory when #70 gets underway 
d) MATLAB pointers are mildly terrifying and the less I have to interact with them, the better

File: tdms/tests/include/abstract_array_test_class.h [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    matlab_input =
+            mxCreateNumericArray(n_dims, (const mwSize *) dimensions, data_type, complex_flag);
+    matlab_input_assigned = true;
+  }
+
+  /* Test functions to be overriden by each class.
+  These methods are designed to be overriden by the subclasses specific to each arrays.h class.
+
+  The default behaviour of each test_ method is to set ran_a_test to false - this means that subclasses that skip over certain functionality (EG do not need to test the result of providing an empty MATLAB input) are skipped in when run_all_class_tests() is run, and do not bloat the log.
+  If overriden, a test_ method should not alter ran_a_test
```
[the remark this answers is not in the document]
Comment (willGraham01) [reply tdms/tests/include/abstract_array_test_class.h:106]: Pretty much. Not all classes need to run every single test, so `AbstractArrayTest` has a private bool `ran_a_test` which is default-initialised to `true`. Then every `test_` function in `AbstractArrayTest` is a virtual method that by default sets `ran_a_test` to `false`.

The `AbstractArrayTest` class then has a method called `run_all_class_tests`, which DWYT. Subclasses of `AbstractArrayTest` can then override the `test_` functions relevant to their class definition, and leave the others alone. `run_all_class_tests` can then be inherited from `AbstractArrayTest` without the need to define it for _every_ subclass, since overridden `test_` methods will then not touch the `ran_a_test` bool, resulting in it remaining `true`, and indicating a test was run.

TLDR; it means that we only have to declare & define overrides for the `test_` methods that are relevant to each class, rather than doing this _plus_ redefining `run_all_class_tests` for every class (plus remembering to update this if we later decide to add more tests to a class).
